redis/credentials.py

```diff
@@ -0,0 +1,77 @@
+from typing import Callable, Optional, Union
+
+
+class CredentialProvider:
+    def __init__(
+        self,
+        username: Union[str, None] = "",
+        password: Union[str, None] = "",
+        supplier: Optional[Callable] = None,
+        *supplier_args,
+        **supplier_kwargs,
+    ):
+        """
+        Initialize a new Credentials Provider.
+        :param supplier: a supplier function that returns the username and password.
+                         def supplier(self, arg1, arg2, ...) -> (username, password)
+                         See examples/connection_examples.ipynb
+        :param supplier_args: arguments to pass to the supplier function
+        :param supplier_kwargs: keyword arguments to pass to the supplier function
+        """
+        self.__username = "" if username is None else username
+        self.__password = "" if password is None else password
```

Comment:
Maybe just a single underscore? I don't think it's a good idea to use Python's private attribute mangling here.

Reply:
@chayim you asked for double underscore, should I revert it?

Reply:
Sorry - these should be single, per side as opposed to dundered, or prefix dunder

Reply:
@chayim - self._username_ ?

Reply:
or `self._username` as we started with

Reply:
single per side == self._username_

Reply:
changed